javascript/computer_science/project_hash_map.md

```diff
@@ -0,0 +1,53 @@
+### Introduction
+
+It is time that you create your own hash map. You already know the magic of how it all works out from the previous lesson, now it is time for you to create your own magical implementation.
+
+#### Self restriction and limitation
+
+  Before we get started, we need to lay down some ground rules. JavaScript's dynamic nature of array allow us to insert and retrieve indexes that is outside our array size range. Example: if we create an array of size `16` to be our buckets size, nothing stopping us from storing items at index `500`. This defeats the purpose we are trying to demonstrate, so we need to put some self restriction to work around this.
+
+  1. When you create a function that takes a hash code value and returns the bucket index, use this snippet before the function returns to make sure it's accessing inbound array indexes only.
+
+      ```javascript
+      if (index < 0 || index >= buckets.length) {
+        throw new Error("Trying to access index out of bound");
+      }
+      ```
+
+  1. Inside the class DO NOT use Objects to store your key value pairs, and DO NOT use the array method `push()` to insert your nodes into the buckets. Always use indexes.
```

Comment:
I think he means using object literals like:
```js
// not okay
{ key: keyToInsert, value: valueToInsert }

// :thumbsup:
new Node(keyToInsert, valueToInsert)  // class
Node(keyToInsert, valueToInsert)  // factory
```
I agree that we need to be precise with the language here. JavaScript mingles the idea of "hash map" and "object" in a way most other languages don't.

Reply:
It could be renamed to Object Literal `{}` and can be more specific on how not to use that object literal, Josh showcase of what is ok and what is not ok is brilliant we can use that

Reply:
It's clearly said in the lesson that key value pairs are stored in the Linked List's node data structure. Then, why do we have to state that objects cannot be used when the learner will always be using a node data structure.